text/0000-io-fs-2.1.md

```diff
@@ -0,0 +1,579 @@
+opt-in to when you're using platform-specific information.
+
+One of the major goals of the `os::unix::fs` design is to enable as much
+functionality as possible when programming against "unix in general" while still
+allowing applications to choose to only program against macos, for example.
+
+#### Fate of `Metadata::{accesed, modified}`
+
+At this time there is no suitable type in the standard library to represent the
+return type of these two functions. The type would either have to be some form
+of time stamp or moment in time, both of which are difficult abstractions to add
+lightly.
+
+Consequently, both of these functions will be **deprecated** in favor of
+requiring platform-specific code to access the modification/access time of
+files. This information is all available via the `MetadataExt` traits listed
+above.
+
+Eventually, once a `std` type for cross-platform timestamps is available, these
+methods will be re-instated as returning that type.
+
+### Lowering and setting `Permissions` (Unix)
+
+> **Note**: this section only describes behavior on unix.
+
+Currently there is no stable method of inspecting the permission bits on a file,
+and it is unclear whether the current unstable methods of doing so,
+`PermissionsExt::mode`, should be stabilized. The main question around this
+piece of functionality is whether to provide a higher level abstractiong (e.g.
+similar to the `bitflags` crate) for the permission bits on unix.
+
+This RFC proposes considering the methods for stabilization as-is and not
+pursuing a higher level abstraction of the unix permission bits. To facilitate
+in their inspection and manipulation, however, the following constants will be
+added:
+
+```rust
+mod os::unix::fs {
+    pub const USER_READ: raw::mode_t;
+    pub const USER_WRITE: raw::mode_t;
+    pub const USER_EXECUTE: raw::mode_t;
+    pub const USER_RWX: raw::mode_t;
+    pub const OTHER_READ: raw::mode_t;
+    pub const OTHER_WRITE: raw::mode_t;
+    pub const OTHER_EXECUTE: raw::mode_t;
+    pub const OTHER_RWX: raw::mode_t;
+    pub const GROUP_READ: raw::mode_t;
+    pub const GROUP_WRITE: raw::mode_t;
+    pub const GROUP_EXECUTE: raw::mode_t;
+    pub const GROUP_RWX: raw::mode_t;
+    pub const ALL_READ: raw::mode_t;
+    pub const ALL_WRITE: raw::mode_t;
+    pub const ALL_EXECUTE: raw::mode_t;
+    pub const ALL_RWX: raw::mode_t;
+}
+```
+
+Finally, the `set_permissions` function of the `std::fs` module is also proposed
+to be marked `#[stable]` soon as a method of blanket setting permissions for a
+file.
+
+## Constructing `Permissions`
+
+Currently there is no method to construct an instance of `Permissions` in a
+cross-platform manner. This RFC proposes adding the following APIs:
+
+```rust
+impl Permissions {
+    /// Creates a new set of permissions appropriate for being placed on a file.
+    ///
+    /// On unix platforms this corresponds to the permission bits `0o666`
```

Comment:
Hm that's a good point! The goal here was to add *some* method of creating a `Permissions` in a cross platform manner, but the exact meaning of the permissions being returned is tough to nail down, and it may mean that we don't want this constructor at all. I've personally only ever needed this when calling `set_permissions` on unix with a known set of bits ahead of time, in which case the `from_mode` constructor below would suffice.

I think I'm somewhat tempted to just remove this constructor altogether for now while we continue to hammer down what `Permissions` means on windows.